### PR TITLE
Add Jemalloc Support to Paddle

### PR DESCRIPTION
**Pull request by sneaxiy**
Jemalloc is a general purpose malloc implementation that emphasizes fragmentation avoidance and scalable concurrency support. It is used to enhance `CPUAllocator` in the future.

### CODE REVIEW

**Comment by gongweibao** [comment]
请教一下几个问题：
1. 为什么是je_malloc而不是tc_malloc?
2. 这个je_malloc可以给我们带来的收益？
3. 静态链接je_malloc会不会有风险？

File: paddle/fluid/memory/jemalloc_test.cc [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+#ifndef _WIN32
+inline void *je_posix_memalign_wrap(size_t alignment, size_t size) {
+  void *ptr = nullptr;
+  je_posix_memalign(&ptr, alignment, size);
+  return ptr;
+}
+
+TEST(jemalloc, je_posix_memalign) {
+  auto x = je_malloc_func(je_posix_memalign_wrap, kAlignment, n - 3);
```
Comment (dzhwinter) [inline paddle/fluid/memory/jemalloc_test.cc:92]: check the address alignment.

File: paddle/fluid/memory/jemalloc_test.cc [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+TEST(jemalloc, je_calloc) {
+  using DataType = float;
+  auto x = je_malloc_func(je_calloc, n, sizeof(DataType));
+  ASSERT_NE(x, nullptr);
+  bool is_all_zero = std::all_of(
+      static_cast<DataType *>(x.get()), static_cast<DataType *>(x.get()) + n,
+      [](const DataType &val) { return val == static_cast<DataType>(0); });
+  ASSERT_TRUE(is_all_zero);
```
Comment (dzhwinter) [inline paddle/fluid/memory/jemalloc_test.cc:59]: it will erase as zero in all platform? Not sure about this test case.

**Comment by dzhwinter** [comment]
Same as above, how about the preload option?  We can change the preload  allocator easily.

**Comment by sneaxiy** [comment]
@gongweibao @dzhwinter 
1. 我们可以对比下jemalloc和tcmalloc的性能，至少从代码上来看TF使用了jemalloc。
2. jemalloc可以用于优化CPU端的内存分配，理论上应该能加速内存分配并减少碎片。
3. 其实静态链接和动态链接理论上都是可行的，TF的做法是在代码里定义了TENSORFLOW_USE_JEMALLOC，作为一个编译选项。

File: paddle/fluid/memory/jemalloc_test.cc (hunk 1 above)
Comment (sneaxiy) [reply paddle/fluid/memory/jemalloc_test.cc:92]: LGTM

File: paddle/fluid/memory/jemalloc_test.cc (hunk 2 above)
Comment (sneaxiy) [reply paddle/fluid/memory/jemalloc_test.cc:59]: `calloc` does guarantee it.

**Comment by gongweibao** [comment]
```
jemalloc可以用于优化CPU端的内存分配，理论上应该能加速内存分配并减少碎片。
```

最好有性能对比。

File: CMakeLists.txt [hunk 3]
```diff
@@ -175,6 +175,7 @@ include(external/eigen)     # download eigen3
 include(external/pybind11)  # download pybind11
 include(external/cares)
 include(external/cub)
+include(external/jemalloc)
```
Comment (velconia) [inline CMakeLists.txt:182]: Maybe ```WITH_JE_MALLOC``` could be a ```cmake``` flags and the allocator could be determined compile time.

Then we could use ```WITH_TC_MALLOC``` and ```WITH_BUDDY_ALLOCATOR``` to change it

**Comment by velconia** [comment]
在CPU模型上大概验证了一下，**在复杂模型（malloc, free特别频繁）的场景下，jemalloc是相比buddy allocator有提升的**，jemalloc能够抹平buddy allocator中因为锁争抢导致的某些时延高峰，但在简单模型（malloc, free不频繁）的场景下，jemalloc的速度相比buddy allocator略有一点点的下降

File: cmake/external/jemalloc.cmake [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+              In Ubuntu, you can use 'apt-get install -y autoconf' to install it.")
+    endif()
+  endif()
+
+  execute_process(COMMAND which autoconf OUTPUT_VARIABLE AUTOCONF_PATH)
+  string(STRIP ${AUTOCONF_PATH} AUTOCONF_PATH)
+  MESSAGE(STATUS "Autoconf: ${AUTOCONF_PATH}")
+
+  set(JEMALLOC_BUILD_COMMAND cd ${JEMALLOC_SOURCES_DIR}/src/extern_jemalloc
+      && ./autogen.sh --with-jemalloc-prefix=je_ --prefix ${JEMALLOC_INSTALL_DIR} && make -j8)
```
Comment (velconia) [inline cmake/external/jemalloc.cmake:52]:
```suggestion
      && EXTRA_CFLAGS=-fPIC EXTRA_CXXFLAGS=-fPIC ./autogen.sh --prefix ${JEMALLOC_INSTALL_DIR} && make -j8
```

File: paddle/fluid/memory/jemalloc_test.cc [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+std::unique_ptr<void, void (*)(void *)> je_malloc_func(MallocFunc &&func,
+                                                       Args... args) {
+  return std::unique_ptr<void, void (*)(void *)>(
+      func(std::forward<Args>(args)...), je_free);
+}
+
+TEST(jemalloc, jemalloc_not_overwrite_std) {
+  ASSERT_TRUE(malloc != je_malloc);
+  ASSERT_TRUE(calloc != je_calloc);
+  ASSERT_TRUE(aligned_alloc != je_aligned_alloc);
```
Comment (velconia) [inline paddle/fluid/memory/jemalloc_test.cc:32]: seems there is no ```aligned_alloc``` in centos

**Pull request closed without merging**